### DISCUSSION TITLE
Unit/Slave ID

### DISCUSSION THREAD

**Discussion by Rid00** [celsworth/lxp-bridge#192] [Q&A]
Hi Celsworth, Reaching out for some help. Firstly thanks for the great work. What Unit_ID / Slave_ID are you using when connecting to the inverter Modbus? For my own personal python project, I would like to try and read the input registers. Your guidance would be appreciated.

**Comment by celsworth** [comment]
Hi :)

I've not seen them referred to as Unit ID / Slave ID before, but I'm guessing you mean what I've called "address". There's actually two in the packets lxp-bridge sends, the first in what seems to be a TCP header and the second in the modbus header.

The one in the TCP header is here:
https://github.com/celsworth/lxp-bridge/blob/c0b781f5a4cd033b15d78d7c3bdbabbd5b00f046/src/lxp/packet.rs#L681
(clearly I didn't know it was an address when I wrote that comment) and it is documented in the first table in https://github.com/celsworth/lxp-bridge/wiki/TCP-Packet-Spec -so this one I set to `1` when writing to the inverter.

The second one is in the modbus header but actually I think I noticed a bug in my wiki documentation while looking this up. In https://github.com/celsworth/lxp-bridge/wiki/TCP-Packet-Spec#tcp-function-0xc2---translated-data I had it as the first byte in the modbus header, but actually I think there's two "length" bytes before it. Added a note to fix it up later.

This  would be set here if it was necessary - https://github.com/celsworth/lxp-bridge/blob/c0b781f5a4cd033b15d78d7c3bdbabbd5b00f046/src/lxp/packet.rs#L950 - but actually since this one is 0 when writing to the inverter, which that array contains by default, its a no-op.

So yeah, the short answer is either 1 or 0, depending :) If you're doing modbus directly over RS485/CAN or something, you won't have a TCP header so I guess you're talking about the one that needs to be 0?

**Reply by Rid00** [reply]
Thank you for the swift reply. I really admire your willingness to help from this comment and others I've seen. From my research, 0 is usually reserved for ID broadcast. So I think the answer I'm looking for is 1. I am trying to access the input registers over TCP using the Pymodbus library in python. I can't believe that i've overlooked your TCP spec wiki. So I think the code I've been using should work over RS485/CAN.

Going to try using your TCP spec sheet to interface with the inverter via the dongle

**Reply by Rid00** [reply]
Chris could you perhaps give me the TCP header and the modbus header you would use to read an input register?

This is what my tcp header looks like : 
`tcp_header = bytes([0xa1, 0x1a, 0x02, 0x00, 0x0c, 0x00, 0x01, 0xc2]) + b'BAXXXXXXXX'`  
#where b'BAXXXXXXXX' is the datalog number in bytes

This is the modbus header: 
```
        crc16 = crcmod.predefined.Crc('crc-16')
        device_address = 0x01
        device_function = 0x04
        inverter_serial = b'\x0D\xD5\x7E\x7D\x00\x00\x00\x00\x00\x00'
        value1 = 20  # Decimal value
        register_low_byte = value1 & 0xFF  # Extract the low byte
        register_high_byte = (value1 >> 8) & 0xFF  # Extract the high byte
        value2 = 5
        num_register_low_byte = value1 & 0xFF  # Extract the low byte
        num_registers_high_byte = (value1 >> 8) & 0xFF  # Extract the high byte

        modbus_payload = bytes([
            device_address,  # Address
            device_function,  # Device Function
            *inverter_serial,  # Inverter Serial Number
            register_low_byte,  # Start address (Low byte)
            register_high_byte,  # Start address (High byte)
            num_register_low_byte,  # Num of registers (Low byte)
            num_registers_high_byte  # Num of registers (High byte)
        ])

        # Calculate the CRC
        crc16.update(modbus_payload)

        # Get the CRC bytes (low byte and high byte)
        crc_low_byte = (crc16.crcValue & 0xFF).to_bytes(1, byteorder='big')
        crc_high_byte = ((crc16.crcValue >> 8) & 0xFF).to_bytes(1, byteorder='big')

        # Add the CRC bytes to the Modbus payload
        modbus_payload += crc_low_byte + crc_high_byte
```

Then the final message sent is : 
`final_header = tcp_header + modbus_payload`

Still I receive no response, except for a heartbeat every few minutes. 
What am I missing :(

**Reply by celsworth** [reply]
I have a test for my TCP packet building which has an example of what a read inputs command should look like:

https://github.com/celsworth/lxp-bridge/blob/c0b781f5a4cd033b15d78d7c3bdbabbd5b00f046/tests/test_packet_tcp_frames.rs#L73-L89

For this test purpose the datalog is `2222222222` and the serial `5555555555`. This is reading registers 0-39 (register start 0, count 40).

So the generated packet in hex is, broken down into digestible chunks:

```
0xa1 0x1a -- static header
0x1 0x0 -- protocol 1
0x20 0x0 -- length, 32 bytes follow
0x1 -- address, 1
0xc2 -- tcp_function = translated data
0x32 0x32 0x32 0x32 0x32 0x32 0x32 0x32 0x32 0x32 -- 2222222222

0x12 0x0 -- length, 18 bytes follow
0x0 -- address, 0
0x4 -- device_function = read inputs
0x35 0x35 0x35 0x35 0x35 0x35 0x35 0x35 0x35 0x35 -- 5555555555
0x0 0x0 -- 0, as in start register
0x28 0x0 -- 40, as in numbers of registers to read
0x2a 0x84 -- checksum
```

I see you're using protocol 2 and my code uses protocol 1. There's some subtle differences, I've never actually tried using protocol 2 for reading inputs though. And your address is 1 whereas I have 0 in the modbus part. Maybe try adjusting those for starters?

If you still get no reply, try bodging in those datalog/serial numbers and verify you get the right checksum (the one above is known to be good for that input packet). If thats right, double and triple check the serial numbers you're sending are right (and the right way round), the inverter won't reply if you don't send the right ones.

And I think my TCP Packet Spec Wiki page needs some cleanup and correction :( Honestly I'm not sure I ever expected anyone to read it!

**Comment by dbainbridge** [comment]
Is the basic idea that you have to send a TCP packet to the dongle and in that packet contains a payload that the firmware on that dongle is interpreting as something that needs to go out over the modbus?

**Reply by celsworth** [reply]
I believe so. This is all internal to the dongle though so I'm not really sure.

I'd be interested if anyone was already on the modbus of the inverter and can see data on it. Perhaps they can confirm then if this is true.

**Comment by deadrabbit87** [comment]
Hi, 

@Rid00 @dbainbridge any sucess here in reading the inverter via modbus TCP?
